### ISSUE TITLE
Singular 4.1.3 reports itself as "version 4.1.2 (4125, 64 bit)"

### ISSUE THREAD

**Issue by zickgraf** [Singular/Sources#987]
When executing `Singular --version` with Singular 4.1.3 (ftp://jim.mathematik.uni-kl.de/pub/Math/Singular/SOURCES/4-1-3/singular-4.1.3.tar.gz) I get the following output:
```
Singular for x86_64-Linux version 4.1.2 (4125, 64 bit) Apr  6 2020 19:20:53 #UNKNOWN_GIT_VERSION
```

I assume the versions in `kernel/mod2.h` are outdated.

**Comment by fingolfin** [comment]
@hannes14 any comments on this?

I really wonder about the process used to make Singular releases. Hans, do you have something like a "release checklist" that you follow step by step to make releases? And/or a script (or multiple) that you use to aid it? My impression is that making a release requires an extraordinary amount of work from you. At the same time, it seems you are the only one who has the knowledge to do it, which may be a serious issue down the road if you ever are unavailable when a release needs to be made.

So I would suggest that we invest some energy into overhauling the Singular release process; it should be automated more, and that what cannot be automated should be documented in a checklist which ideally would be accessible to the general public (e.g. as a file in the repo or in a Wiki or something like that). Perhaps this is already the case, though -- in that case, I'd be interested to learn where I can find this!

CC @wdecker @fieker

**Comment by zickgraf** [comment]
This seems to be fixed by 161411bdc5cc30f7965d2c36a5250edcb33d293e in version 4.1.3p2. @fingolfin should I close this issue or do you want to keep this open for the discussion of the points you raised?

**Comment by orlitzky** [comment]
I think the benefits of having the version declared in a single place are clear? The `AC_INIT` macro in `configure.ac` already defines a `PACKAGE_VERSION` constant that is used elsewhere. That would be my first choice for a blessed location, as it can directly replace `SINGULAR_VERSION`.

The file `kernel/mod2.h` goes to some trouble to define `SINGULAR_MAJOR_VERSION` et al. individually, but it's not clear how/if those constants are used. Using pkg-config (or the soname) would be a more modern way to detect API/ABI compatibility.

**Comment by fingolfin** [comment]
@orlitzky I agree that in general, declaring the version in as few places as possible, ideally just one is best. However, it is not always practical to do it in just one place. However, if one uses a script to update the version, it doesn't even matter in how many places it is.

`SINGULAR_MAJOR_VERSION` is useful for code which is compiled against libsingular, e.g. to compare compile time and runtime versions of Singular, or to enable/disable features. This is completely orthogonal to `pkg-config`.

Anyway, I've submitted issue #1023 now, which would help avoid this kind of problem in the future. @zickgraf I think this issue can be closed now.

**Comment by orlitzky** [comment]
Maybe linearly independent, but certainly not orthogonal =)

There are two main things a consumer wants to know about libsingular:

1. libsingular was updated, do I need to be rebuilt? The library soname is more appropriate than the package version  for this.
2. I'm about to be built, can I enable some feature that's only available in newer versions of libsingular? For this pkg-config works fine, since it can easily test the major version number.

**Comment by fingolfin** [comment]
Regarding 2, I think pkg-config is a bit of red herring there; what you seem to be really saying is that if one has a `configure` script, then one can use that to detect the version of Singular (via pkg-config or some other means, doesn't matter), and then use that to `#define` macros to control which parts of your program are compiled. And that's fair, but for a simple program, it is often much more convenient to check the existing macro `SINGULAR_MAJOR_VERSION` (and its companions).

And sometimes one really just wants to be able to access the version of a library the code was built against; e.g. to print out "Built against Singular X.Y, linked against Singular A.B" -- that can be tremendously helpful in debugging.

In summary: just because you don't see an application for a feature doesn't mean there isn't one :-). And even if there was none: you can't just remove a `#define` that existing code out there relies, unless you really like breaking other people's code for no good reasons, and the consequences of it...

But this is all really just a distraction from the actual issue: that we should make sure that we can't forget for a release to update the Singular version in all places it needs to updated. How we achieve that, either by reducing the place the version occurs in, or by scripting, or any other means, is less important.

**Comment by orlitzky** [comment]
The benefit of using pkg-config even for a simple program is that I don't have to know how to compile that program against libsingular before I know if I can compile it against libsingular. Instead the `*.pc` file makes it available through a standard interface, `pkg-config --modversion Singular`, that doesn't depend on any knowledge beyond the name "Singular".

In any case, I'm not saying don't leave it in there, just providing ammunition if someone wants to consider it. Computing the "whole" version in `configure.ac` from three separate major/minor/patch versions (which would then supply e.g. `SINGULAR_MAJOR_VERSION`) should be possible as well. Whatever works.
